Add docopt recipe

r0man opened the pull request:
### Brief summary of what the package does

The docopt.el Emacs package is a Docopt (http://docopt.org)
implementation in Emacs Lisp. It has functionality to parse a Docopt
usage/help string and provide an Emacs transient command to execute
the Docopt program, with the user provided inputs.

At my current work we have a lot of command line tools that people
use. These tools all use Docopt as the help format and I'm using this
package to build a magit like interface on top of them.

### Direct link to the package repository

https://github.com/r0man/docopt.el

### Your association with the package

I'm the maintainer.

### Relevant communications with the upstream package maintainer

**None needed**

### Checklist

Please confirm with `x`:

- [x] The package is released under a [GPL-Compatible Free Software License](https://www.gnu.org/licenses/license-list.en.html#GPLCompatibleLicenses).
- [x] I've read [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)
- [x] I've used the latest version of [package-lint](https://github.com/purcell/package-lint) to check for packaging issues, and addressed its feedback
- [?] My elisp byte-compiles cleanly
- [?] `M-x checkdoc` is happy with my docstrings
- [x] I've built and installed the package using the instructions in [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)
- [ ] I have confirmed some of these without doing them

#### Byte compile errors

I tried to remove most byte compile errors. Those are the ones I could
not figure out yet.

```
[roman@thinkpad docopt.el]$ make v=vvv lint-compile
LOG (2020-10-23 13:00:02): Linting compilation...
Package cl is deprecated
Error loading autoloads: (void-function helm-make-source)

In toplevel form:
src/docopt-argv.el:103:1:Error: variable ‘_’ not left unused

In toplevel form:
src/docopt-option.el:312:1:Error: the following functions are not known to be defined: option-1, option-2

In toplevel form:
src/docopt-parser.el:187:1:Error: variable ‘_’ not left unused

In toplevel form:
src/docopt-program.el:250:1:Error: the function ‘docopt-parse’ is not known to be defined.

In toplevel form:
src/docopt-transient.el:333:1:Error: Unused lexical variable ‘vterm-shell’
ERROR (2020-10-23 13:00:04): Linting compilation failed.
LOG (2020-10-23 13:00:04): Finished with 1 errors.
make: *** [Makefile:52: lint-compile] Error 1
```

#### Checkdoc Errors

I think I removed all checkdoc errors, except the ones releated to
aspell. Any ideas I can tell checkdoc to not warn about those words?

```
[roman@thinkpad docopt.el]$ make v=vvv lint-checkdoc
LOG (2020-10-23 13:03:04): Linting checkdoc...
Package cl is deprecated
Error loading autoloads: (void-function helm-make-source)
Starting new Ispell process /usr/bin/aspell with default dictionary... \
Starting new Ispell process /usr/bin/aspell with default dictionary...done
docopt-analyzer.el:48: Error checking word DOCOPT using aspell with default dictionary
docopt-analyzer.el:103: Error checking word DEDUPLICATE using aspell with default dictionary
docopt-analyzer.el:103: Error checking word DOCOPT using aspell with default dictionary
docopt-analyzer.el:195: Error checking word DOCOPT using aspell with default dictionary
docopt-argument.el:73: Error checking word EQUAL-ISH using aspell with default dictionary
docopt-argument.el:79: Error checking word DOCOPT using aspell with default dictionary
docopt-argument.el:87: Error checking word DOCOPT using aspell with default dictionary
docopt-argument.el:98: Error checking word DOCOPT using aspell with default dictionary
docopt-argv.el:240: All variables and subroutines might as well have a documentation string
docopt-command.el:69: Error checking word DOCOPT using aspell with default dictionary
docopt-command.el:72: Error checking word DOCOPT using aspell with default dictionary
docopt-command.el:81: Error checking word DOCOPT using aspell with default dictionary
docopt-either.el:64: Error checking word DOCOPT using aspell with default dictionary
docopt-either.el:74: Error checking word DOCOPT using aspell with default dictionary
docopt-either.el:79: Error checking word DOCOPT using aspell with default dictionary
docopt-either.el:97: Error checking word DOCOPT using aspell with default dictionary
docopt-either.el:101: Error checking word DOCOPT using aspell with default dictionary
docopt-either.el:105: Error checking word DOCOPT using aspell with default dictionary
docopt-generic.el:51: Error checking word DOCOPT using aspell with default dictionary
docopt-generic.el:54: Error checking word DOCOPT using aspell with default dictionary
docopt-generic.el:58: Error checking word DOCOPT using aspell with default dictionary
docopt-generic.el:65: Error checking word DOCOPT using aspell with default dictionary
docopt-generic.el:69: Error checking word DOCOPT using aspell with default dictionary
docopt-generic.el:72: Error checking word DOCOPT using aspell with default dictionary
docopt-generic.el:76: Error checking word EQUAL-ISH using aspell with default dictionary
docopt-generic.el:79: Error checking word EQUAL-ISH using aspell with default dictionary
docopt-generic.el:83: Error checking word DOCOPT using aspell with default dictionary
docopt-generic.el:90: Error checking word DOCOPT using aspell with default dictionary
docopt-generic.el:104: Error checking word DOCOPT using aspell with default dictionary
docopt-group.el:74: Error checking word DOCOPT using aspell with default dictionary
docopt-group.el:85: Error checking word DOCOPT using aspell with default dictionary
docopt-group.el:90: Error checking word DOCOPT using aspell with default dictionary
docopt-group.el:100: Error checking word DOCOPT using aspell with default dictionary
docopt-group.el:104: Error checking word DOCOPT using aspell with default dictionary
docopt-group.el:108: Error checking word DOCOPT using aspell with default dictionary
docopt-group.el:112: Error checking word DOCOPT using aspell with default dictionary
docopt-group.el:116: Error checking word DOCOPT using aspell with default dictionary
docopt-group.el:121: Error checking word DOCOPT using aspell with default dictionary
docopt-option.el:87: Error checking word EQUAL-ISH using aspell with default dictionary
docopt-option.el:137: Error checking word DOCOPT using aspell with default dictionary
docopt-option.el:147: Error checking word DOCOPT using aspell with default dictionary
docopt-option.el:151: Error checking word DOCOPT using aspell with default dictionary
docopt-option.el:156: Error checking word DOCOPT using aspell with default dictionary
docopt-option.el:191: Error checking word DOCOPT using aspell with default dictionary
docopt-option.el:198: Error checking word DOCOPT using aspell with default dictionary
docopt-option.el:202: Error checking word DOCOPT using aspell with default dictionary
docopt-option.el:251: Error checking word DOCOPT using aspell with default dictionary
docopt-options-shortcut.el:62: Error checking word DOCOPT using aspell with default dictionary
docopt-options-shortcut.el:66: Error checking word DOCOPT using aspell with default dictionary
docopt-options-shortcut.el:70: Error checking word DOCOPT using aspell with default dictionary
docopt-options-shortcut.el:74: Error checking word DOCOPT using aspell with default dictionary
docopt-options-shortcut.el:85: Error checking word DOCOPT using aspell with default dictionary
docopt-parser.el:81: Error checking word DOCOPT using aspell with default dictionary
docopt-parser.el:89: Error checking word DOCOPT using aspell with default dictionary
docopt-parser.el:242: Error checking word DOCOPT using aspell with default dictionary
docopt-parser.el:252: Error checking word DOCOPT using aspell with default dictionary
docopt-parser.el:258: Error checking word DOCOPT using aspell with default dictionary
docopt-parser.el:264: Error checking word DOCOPT using aspell with default dictionary
docopt-parser.el:324: Error checking word DOCOPT using aspell with default dictionary
docopt-parser.el:421: Argument ‘state’ should appear (as STATE) in the doc string
docopt-program.el:39: Error checking word DOCOPT using aspell with default dictionary
docopt-program.el:89: Error checking word DOCOPT using aspell with default dictionary
docopt-program.el:93: Error checking word DOCOPT using aspell with default dictionary
docopt-program.el:97: Error checking word DOCOPT using aspell with default dictionary
docopt-program.el:104: Error checking word DOCOPT using aspell with default dictionary
docopt-program.el:130: Error checking word EQUAL-ISH using aspell with default dictionary
docopt-program.el:168: Error checking word DOCOPT using aspell with default dictionary
docopt-program.el:180: Error checking word DOCOPT using aspell with default dictionary
docopt-program.el:191: Error checking word DOCOPT using aspell with default dictionary
docopt-program.el:209: Error checking word DOCOPT using aspell with default dictionary
docopt-program.el:214: Error checking word DOCOPT using aspell with default dictionary
docopt-program.el:218: Error checking word DOCOPT using aspell with default dictionary
docopt-program.el:222: Error checking word DOCOPT using aspell with default dictionary
docopt-program.el:226: Error checking word DOCOPT using aspell with default dictionary
docopt-program.el:233: Error checking word DOCOPT using aspell with default dictionary
docopt-repeated.el:43: Error checking word DOCOPT using aspell with default dictionary
docopt-repeated.el:52: Error checking word DOCOPT using aspell with default dictionary
docopt-repeated.el:56: Error checking word DOCOPT using aspell with default dictionary
docopt-repeated.el:60: Error checking word DOCOPT using aspell with default dictionary
docopt-repeated.el:64: Error checking word DOCOPT using aspell with default dictionary
docopt-repeated.el:68: Error checking word DOCOPT using aspell with default dictionary
docopt-standard-input.el:38: Error checking word DOCOPT using aspell with default dictionary
docopt-standard-input.el:42: Error checking word DOCOPT using aspell with default dictionary
docopt-standard-input.el:46: Error checking word DOCOPT using aspell with default dictionary
docopt-standard-input.el:54: Error checking word DOCOPT using aspell with default dictionary
docopt-transient.el:44: Error checking word DOCOPT using aspell with default dictionary
docopt-transient.el:118: Error checking word SHORTARG using aspell with default dictionary
docopt-transient.el:132: Error checking word DOCOPT using aspell with default dictionary
docopt-transient.el:147: Error checking word DOCOPT using aspell with default dictionary
docopt-transient.el:165: Error checking word DOCOPT using aspell with default dictionary
docopt-transient.el:182: Error checking word DOCOPT using aspell with default dictionary
docopt-transient.el:253: Error checking word S-EXPRS using aspell with default dictionary
docopt-transient.el:259: Error checking word S-EXPRS using aspell with default dictionary
docopt-transient.el:266: Error checking word S-EXPRS using aspell with default dictionary
docopt-transient.el:275: Error checking word S-EXPRS using aspell with default dictionary
docopt-transient.el:280: Error checking word S-EXPRS using aspell with default dictionary
docopt-transient.el:285: Error checking word S-EXPRS using aspell with default dictionary
docopt-transient.el:290: Error checking word S-EXPRS using aspell with default dictionary
docopt-transient.el:355: Error checking word DOCOPT using aspell with default dictionary
docopt-transient.el:366: Error checking word DOCOPT using aspell with default dictionary
docopt-transient.el:377: Error checking word DOCOPT using aspell with default dictionary
docopt-transient.el:389: Error checking word DOCOPT using aspell with default dictionary
docopt-transient.el:437: Error checking word DOCOPT using aspell with default dictionary
docopt-transient.el:476: Error checking word S-EXPRS using aspell with default dictionary
docopt-transient.el:492: Error checking word S-EXPRS using aspell with default dictionary
docopt-usage-pattern.el:74: Error checking word DOCOPT using aspell with default dictionary
docopt-usage-pattern.el:79: Error checking word DOCOPT using aspell with default dictionary
docopt-usage-pattern.el:91: Error checking word DOCOPT using aspell with default dictionary
docopt-usage-pattern.el:95: Error checking word DOCOPT using aspell with default dictionary
docopt-usage-pattern.el:99: Error checking word DOCOPT using aspell with default dictionary
docopt-usage-pattern.el:103: Error checking word DOCOPT using aspell with default dictionary
docopt-util.el:59: Error checking word DOCOPT using aspell with default dictionary
docopt.el:44: Error checking word DOCOPT using aspell with default dictionary
docopt.el:50: Error checking word DOCOPT using aspell with default dictionary
docopt.el:59: Error checking word DOCOPT using aspell with default dictionary
docopt.el:64: Error checking word DOCOPT using aspell with default dictionary
```

alphapapa commented:
This looks like a very useful package.  Here's some feedback:

1.  That's a *lot* of source files.  That's not necessarily a problem, but it's unusual for an Emacs package to have so many.  You might consider consolidating some of them.
2.  That might be made easier by eliminating some code that doesn't seem to be necessary.  For example, in `docopt-util.el` there are several functions that seem to be covered by functions in Emacs, `cl-lib.el`, and/or `dash.el`.  For example, `docopt-substring`, `docopt-strip`, `docopt-find`.  And several appear trivial and should probably be inlined, like `docopt-remove-duplicates`, `docopt-bold`, `docopt-keyword`.  As well, some other files seem to have functions that should be covered by other libraries, like `docopt-walk`.
3.  The `-pkg.el` file isn't forbidden, but it isn't necessary to have one, and MELPA generally prefers that packages don't use them.  Again, if you consolidate the source code, you might find that this isn't necessary.
4.  In `docopt-argv--parse-simple-list`, you use `eval` on a backquoted list returned by another function.  Instead, you should probably use a lambda with a `funcall`.  Considering how much of `seq`, `dash`, and mapping functions you use elsewhere, I think that should be fairly natural here.
5.  `Package cl is deprecated`  Use `cl-lib`.  See the `cl` Info manual for more information.
6.  `Error loading autoloads: (void-function helm-make-source)`  Generally you shouldn't include Helm-specific code in your package.  If you want to support Helm, you should generally put it in a separate `helm-*` package (although I'm guilty of violating that rule in some of my own packages), or you can very carefully use `with-eval-after-load`, although that's not recommended (I do that in some of mine, but it's fragile and tends to still cause warnings like this).
7.  `Error: variable ‘_’ not left unused`  In Elisp, naming a variable with the prefix `_` indicates to the byte-compiler that its value is not intended to be used.  So either you used such a variable name by accident, in which case you can simply give it a meaningful name, or it might be caused by something else, like using `eval`.
8.  `src/docopt-program.el:250:1:Error: the function ‘docopt-parse’ is not known to be defined.` This is one of the downsides to using so many source files: it's easy to forget to `require` functions from a different file, or to end up with circular dependencies.
9.  `src/docopt-transient.el:333:1:Error: Unused lexical variable ‘vterm-shell’` I haven't looked at the code, but the error means what it says.  Since that symbol isn't from your package, it's probably caused by a macro you're calling.
10.  I don't know why Checkdoc is giving those warnings.  Try using `makem.sh`'s sandbox feature to ensure that your personal Emacs config isn't causing it.

r0man commented:
Thanks for the review @alphapapa, I'm looking into your suggestions.

r0man commented:
Hi @alphapapa,

some of the errors came from not running the linters in a sandbox, the
warning about `helm-make-source`, `Package cl is deprecated` for
example.

I fixed the circular dependency problem, removed the `docopt-pkg.el`
file and got rid of all the other warnings, except for the following 2
warnings:

The first one is this checkdoc warning:

```
docopt-argv.el:243: All variables and subroutines might as well have a documentation string
```

I added a docstring [here](https://github.com/r0man/docopt.el/blob/master/src/docopt-argv.el#L243),
but checkdoc is still complaining. Do you have any idea how to solve this?

The second one is about [emacs-vterm](https://github.com/akermu/emacs-libvterm).

```
src/docopt-transient.el:334:1:Error: Unused lexical variable ‘vterm-shell’
```

Since emacs-vterm uses Emacs modules and it's a bit more complicated
to install, I wanted to have this as an optional dependency and not
require it if users don't have it already installed. It falls back to
using a regular term if not installed.

I tried all sorts of things, like protecting my code with a `(fboundp 'vterm-shell)`, 
using `(ignore vterm-shell)`, but I could not get rid of this warning.

I also tried making it a hard dependency, but run into issues
installing emacs-libvterm on the CI box.

Regarding the number of files, and the helper functions, I would like
to keep them as they are for now. The circular dependency issue is
solved, and the functions in docopt-util.el are a bit more specialized
than what I found in other libraries.

How can we move forward?

Thanks, Roman.

riscy commented:
Regarding this:
```
Warning (emacs): 
docopt-argv.el:243: All variables and subroutines might as well have a documentation string
```
It looks like checkdoc doesn't recognize the qualifier so this is a false positive.

The vterm stuff is trickier -- sometimes a cleaner way to handle a dependency you want to make optional is to create an add-on package that enables the feature.  Otherwise for these warnings:
```
docopt-transient.el:334:1:Warning: Unused lexical variable `vterm-shell'
docopt-transient.el:334:1:Warning: Unused lexical variable
    `vterm-kill-buffer-on-exit'
In docopt-transient--execute-command-vterm:
docopt-transient.el:343:27:Warning: reference to free variable
    `vterm--process'
docopt-transient.el:346:20:Warning: reference to free variable
    `vterm-copy-mode-map'
In end of data:
docopt-transient.el:532:1:Warning: the function `vterm-mode' is not known to
    be defined.
```
you can use e.g. `(defvar vterm--process)` at the top of the file to declare a variable's existence but not set it to anything, and I believe this will also work to silence the "unused lexical variable" warnings.  You can use `(declare-function ... "ext:vterm.el")` to declare that vterm-mode exists in some external file.

I agree with @alphapapa that there are a large number of files averaging a couple hundred lines per file (with 30 lines of boilerplate each), and also strongly agree with inlining where it makes sense.  These issues aren't blockers but these patterns are unusual for elisp, mostly due to community style but also due to a lack of tooling (you saw how easy it was to create a circular dependency) and the fact that it often isn't as necessary to follow a hypermodularized "enterprise" style.

Minor:
- docopt-program.el#L234: Can the `format` and `concat` be combined?

alphapapa commented:
@r0man FYI:

* https://github.com/r0man/docopt.el/commit/f819f57cbec242e49bfc84be4532ae7e71f539fa#r43994453

r0man commented:
Hi @riscy,

thanks for reviewing this. Thanks for your tip about the `(defvar
vterm--process)`, I finally got rid of the vterm warning. :)

Regarding the checkdoc false positive, is there a way to disable
checkdoc just for this function, so I can run the linter on CI and
fail if a new warning gets introduced?

Regarding your comment about combining `format` and `concat`, I don't
think so, because the first `concat` builds the format string and the
other one is for each of the arguments.

Some other things I did:

- I removed some files in https://github.com/r0man/docopt.el/pull/19

- I replaced seq-let with pcase-let https://github.com/r0man/docopt.el/pull/18  //cc @alphapapa

- I removed/inlined docopt-strip

- I removed/inlined docopt-remove-duplicates

Roman

riscy commented:
> is there a way to disable checkdoc just for this function, so I can run the linter on CI and fail if a new warning gets introduced?

Not to my knowledge unfortunately -- I agree it's useful being able to run the linter without false positives (in my opinion it's worth bending the code to make them happy).  This looks ready to merge, though. :)

r0man commented:
Nice. Thank you all!

alphapapa commented:
FYI, I took the liberty of posting it to Reddit, because this is too cool!  https://www.reddit.com/r/emacs/comments/jux8c1/ann_r0mandocoptel_run_docoptcompatible_shell/
